### ISSUE TITLE
Rounding errors when computing inertia in KMeans

### ISSUE THREAD

**Issue by fcharras** [soda-inria/sklearn-numba-dpex#28]
This issue to track a point brought by @jeremiedbb regarding computing inertia: computing (x-y)2 is more numerically stable than computing x2 + 2*(-xy + y2/2) and where inertia is printed or returned to the user, the numerical stability should be preferred. Our `get_inertia` and `get_labels_and_inertia` kernels should be amended to use (x-y)2

**Comment by jjerphan** [comment]
The current implementation is not the stablest indeed, but should be the most computational intensive one.

It the use of this decomposition currently causing problem for the implementation?

**Comment by jeremiedbb** [comment]
Here's an example where it can fail hard
```py
import numpy as np
from sklearn.cluster import KMeans

X = np.array([[-1.0001],[-0.9999],[0.9999],[1.0001]], dtype=np.float32)
km = KMeans(n_clusters=2, n_init=1, algorithm='full')
km.fit(X)

# The centroids are correct: it's fine to use the x2 + 2*(-xy + y2/2) formula to find the labels
km.cluster_centers_
>>> [[ 1.]
     [-1.]]

# using the x2 + 2*(-xy + y2/2)  [note that this is not from the actual sklearn KMeans :) ]
km.inertia_
>>> 0.0

# correct inertia
((X - km.cluster_centers_[km.labels_])**2).sum()
>>> 4.0013276e-08
```

**Comment by fcharras** [comment]
TY for the reproducing example !

`4e-08` error doesn't sound too bad for float32 compute ? let's keep this issue open and maybe come back at it in later stages.

I think the performance impact of returning the accurate inertia in score or fit method will be unnoticeable, I'm less sure about using that when `verbose=True`

**Comment by fcharras** [comment]
OK. I'll issue a PR to fix this. A side point is that fusing computations of inertia and labels is counter productive, that explains the choices in sklearn that we didn't understand.
